R-package/R/callbacks.R

```diff
@@ -1,563 +1,923 @@
-
-  callback <- function(env = parent.frame()) {
-    if (length(env$bst_evaluation) == 0 ||
-        period == 0 ||
-        NVL(env$rank, 0) != 0)
-      return()
-
-    i <- env$iteration
-    if ((i - 1) %% period == 0 ||
-        i == env$begin_iteration ||
-        i == env$end_iteration) {
-      stdev <- if (showsd) env$bst_evaluation_err else NULL
-      msg <- .format_eval_string(i, env$bst_evaluation, stdev)
-      cat(msg, '\n')
-    }
-  }
-  attr(callback, 'call') <- match.call()
-  attr(callback, 'name') <- 'cb.print.evaluation'
-  callback
-}
-
-
-#' Callback closure for logging the evaluation history
+#' For \link{xgb.cv}, folds are a list with a structure as follows:\itemize{
+#' \item `dtrain`: The training data for the fold (as an `xgb.DMatrix` object).
+#' \item `bst`: Rhe `xgb.Booster` object for the fold.
+#' \item `watchlist`: A list with two DMatrices, with names `train` and `test`
+#' (`test` is the held-out data for the fold).
+#' \item `index`: The indices of the hold-out data for that fold (base-1 indexing),
+#' from which the `test` entry in the watchlist was obtained.
+#' }
 #'
-#' @details
-#' This callback function appends the current iteration evaluation results \code{bst_evaluation}
-#' available in the calling parent frame to the \code{evaluation_log} list in a calling frame.
+#' This object should \bold{not} be in-place modified in ways that conflict with the
+#' training (e.g. resetting the parameters for a training update in a way that resets
+#' the number of rounds to zero in order to overwrite rounds).
 #'
-#' The finalizer callback (called with \code{finalize = TURE} in the end) converts
-#' the \code{evaluation_log} list into a final data.table.
+#' Note that any R attributes that are assigned to the booster during the callback functions,
+#' will not be kept thereafter as the booster object variable is not re-assigned during
+#' training. It is however possible to set C-level attributes of the booster through
+#' \link{xgb.attr} or \link{xgb.attributes}, which should remain available for the rest
+#' of the iterations and after the training is done.
 #'
-#' The iteration evaluation result \code{bst_evaluation} must be a named numeric vector.
+#' For keeping variables across iterations, it's recommended to use `env` instead.
+#' \item data The data to which the model is being fit, as an `xgb.DMatrix` object.
 #'
-#' Note: in the column names of the final data.table, the dash '-' character is replaced with
-#' the underscore '_' in order to make the column names more like regular R identifiers.
+#' Note that, for \link{xgb.cv}, this will be the full data, while data for the specific
+#' folds can be found in the `model` object.
+#'
+#' \item watchlist The evaluation watchlist, as passed under argument `watchlist` to
+#' \link{xgb.train}.
+#'
+#' For \link{xgb.cv}, this will always be `NULL`.
+#'
+#' \item begin_iteration Index of the first boosting iteration that will be executed
+#' (base-1 indexing).
```

Comment:
I think it should always be 1 and `num_boost_rounds` should always be the number of *new* trees to add. This helps XGBoost to provide additional checks during the first iteration with new data, and fewer conditions are easier to understand.

Reply:
There's also a functionality that updates trees without adding new ones.

Reply:
they can get it from the model itself. let's keep this part consistent across language bindings.

Reply:
The python interface also passes it, as a parameter named `epoch`.

For example, this is the current python signature for the equivalent function to run before each round:
```python
def before_iteration(
        self,
        model: _Model,
        epoch: int,
        dtrain: DMatrix,
        evals: Optional[List[Tuple[DMatrix, str]]],
    ) -> bool
```

Reply:
> This will typically be '1', but when using training continuation, depending on the
#' parameters for updates

It's passed in Python, but I think it should start from 0/1 instead of the previous boosted number.

Reply:
But imagine again a case in which a user is setting up a custom step size sequence: if this were to be reset to zero on training continuation, then the user callback would give out different results when training in one go than when using continuation.

Reply:
That's a good point, we might want to break the compatibility in other language bindings for this change.